### PR TITLE
Telemetry: wrap heartbeat struct in NSObject; handle speech manager edge cases

### PR DESCRIPTION
**Pull request by andymoe**
- Fix logic error with arming status voice/ui update
- Nil out craft heartbeat object so labels are sure to be
  populated and status spoken when app enters the
  foreground again.
- Handle case where telemetry is lost; nil out craft heartbeat object etc.

### CODE REVIEW

**Comment by andymoe** [comment]
Hey @cnatoli in order to fix the armed status bug I had to go all in with the heartbeat wrapping ticket and since there were so many edge cases that needed addressed. I also had to do a fair amount of refactoring in spots to make things work and not be a huge hack job. If you have any time tonight for a quick pass I'd really appreciate it.

@btgalush I could use your eyes as well. I've done as much testing here as I can without the RC controller to arm the copter and flip though modes.

File: iGCS/Models/Craft/GCSCraftArduCopter.m [hunk 1]
```diff
@@ -6,12 +6,19 @@
...
+#import "TelemetryManagers.h"
 #import "GCSCraftArduCopter.h"
 #import "GCSCraftNotifications.h"
 #import "MavLinkUtility.h"
 #import "GCSCraftMixins.h"
 #import "Mixin.h"
 
+
+@interface GCSCraftArduCopter ()
+@property (nonatomic, strong) NSNotificationCenter *notificationCenter;
```
Comment (cnatoli) [inline iGCS/Models/Craft/GCSCraftArduCopter.m]: DRY - move to base class.

File: iGCS/Models/Craft/GCSCraftArduCopter.m [hunk 2]
```diff
@@ -30,11 +37,34 @@ @implementation GCSCraftArduCopter
...
+                                             [weakSelf clearHearbeat];
+                                         }];
+
+        // set heartbeat to nil when app loses telemetry for a period of time
+        [self.notificationCenter addObserverForName:GCSHeartbeatManagerHeartbeatWasLost
+                                             object:nil
+                                              queue:self.craftQueue
+                                         usingBlock:^(NSNotification *note) {
+                                             [weakSelf clearHearbeat];
+                                         }];
```
Comment (cnatoli) [inline iGCS/Models/Craft/GCSCraftArduCopter.m]: DRY

File: iGCS/Models/Craft/GCSCraftArduCopter.m [hunk 3]
```diff
@@ -30,11 +37,34 @@ @implementation GCSCraftArduCopter
...
     }
 
     return self;
 }
 
+- (void)dealloc {
```
Comment (cnatoli) [inline iGCS/Models/Craft/GCSCraftArduCopter.m]: Ok, will stop with the DRYs. We have a class hierarchy here, so let's express it as one (especially before we end up with additional craft models)

File: iGCS/Models/Craft/GCSCraftNotifications.h [hunk 4]
```diff
@@ -7,15 +7,15 @@
...
 #import <Foundation/Foundation.h>
-#import "mavlink.h"
+#import "GCSHeartbeat.h"
 
 @interface GCSCraftNotifications : NSObject
 
-+ (void)didNavModeChangeFromLastHeartbeat:(mavlink_heartbeat_t) lastHeartbeat
-                      andNewHeartbeat:(mavlink_heartbeat_t) newHeartbeat;
++ (void)didNavModeChangeFromLastHeartbeat:(GCSHeartbeat *) lastHeartbeat
+                      andNewHeartbeat:(GCSHeartbeat *) newHeartbeat;
```
Comment (cnatoli) [inline iGCS/Models/Craft/GCSCraftNotifications.h]: Indentation.

File: iGCS/Models/Craft/GCSCraftNotifications.m [hunk 5]
```diff
@@ -14,10 +14,11 @@
 
 @implementation GCSCraftNotifications
 
-+ (void)didNavModeChangeFromLastHeartbeat:(mavlink_heartbeat_t) lastHeartbeat
-                      andNewHeartbeat:(mavlink_heartbeat_t) newHeartbeat {
++ (void)didNavModeChangeFromLastHeartbeat:(GCSHeartbeat *) lastHeartbeat
+                      andNewHeartbeat:(GCSHeartbeat *) newHeartbeat {
```
Comment (cnatoli) [inline iGCS/Models/Craft/GCSCraftNotifications.m]: Indentation

File: iGCS/Models/Craft/GCSCraftNotifications.m [hunk 6]
```diff
@@ -25,11 +26,12 @@ + (void)didNavModeChangeFromLastHeartbeat:(mavlink_heartbeat_t) lastHeartbeat
...
-+ (void)didArmedStatusChangeFromLastHeartbeat:(mavlink_heartbeat_t) lastHeartbeat
-                              andNewHeartbeat:(mavlink_heartbeat_t) newHeartbeat {
++ (void)didArmedStatusChangeFromLastHeartbeat:(GCSHeartbeat *) lastHeartbeat
+                              andNewHeartbeat:(GCSHeartbeat *) newHeartbeat {
+
 
-    if (lastHeartbeat.base_mode != 0 &&
-        (lastHeartbeat.base_mode & MAV_MODE_FLAG_SAFETY_ARMED) == (newHeartbeat.base_mode & MAV_MODE_FLAG_SAFETY_ARMED)) {
+    // if lastHeartbeat is not nil AND the armings status has not changes return
+    if (lastHeartbeat && (lastHeartbeat.isArmed) == (newHeartbeat.isArmed)) {
```
Comment (cnatoli) [inline iGCS/Models/Craft/GCSCraftNotifications.m]: No need for comment, or additional parens,

File: iGCS/Models/Telemetry/GCSHeartbeat.m [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+@interface GCSHeartbeat ()
+@property (nonatomic, assign) mavlink_heartbeat_t heartbeat;
+- (instancetype)initWithHeartbeat:(mavlink_heartbeat_t) heartbeat;
+@end
+
+@implementation GCSHeartbeat
+
++ (GCSHeartbeat *)heartbeatFromMavlinkHeartbeat:(mavlink_heartbeat_t)heartbeat {
+    GCSHeartbeat *aHeartbeat = [[GCSHeartbeat alloc] initWithHeartbeat:heartbeat];
+    return aHeartbeat;
```
Comment (cnatoli) [inline iGCS/Models/Telemetry/GCSHeartbeat.m]: No need for local var

File: iGCS/Managers/Telemetry/GCSHeartbeatManager.m [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+//
+
+#import "GCSHeartbeatManager.h"
+
+NSString * const GCSHeartbeatManagerHeartbeatWasLost = @"GCSHeartbeatManagerHeartbeatWasLost";
+NSString * const GCSHeartbeatManagerHeartbeatDidArrive = @"GCSHeartbeatManagerHeartbeatDidArrive";
+
+@implementation GCSHeartbeatManager
+
+static const double HEARTBEAT_LOSS_WAIT_TIME = 3.0;
```
Comment (cnatoli) [inline iGCS/Managers/Telemetry/GCSHeartbeatManager.m]: Remove similar const in GCSMapViewController

File: iGCS/UserInterface/ViewControllers/GCSMapViewController.m [hunk 9]
```diff
@@ -449,14 +485,10 @@ - (void) handlePacket:(mavlink_message_t*)msg {
         case MAVLINK_MSG_ID_HEARTBEAT: {
             mavlink_heartbeat_t heartbeat;
             mavlink_msg_heartbeat_decode(msg, &heartbeat);
-            
-            // We got a heartbeat, so...
-            [self rescheduleHeartbeatLossCheck];
```
Comment (cnatoli) [inline iGCS/UserInterface/ViewControllers/GCSMapViewController.m]: Remove unused method rescheduleHeartbeatLossCheck

File: iGCS/UserInterface/ViewControllers/GCSMapViewController.m [hunk 10]
```diff
@@ -184,6 +204,22 @@ - (void)craftCustomModeDidChange {
     [self.customModeLabel setText:[GCSDataManager sharedInstance].craft.currentModeName];
 }
 
+- (void)resetTelemetryUserInterface {
+    [self.armedLabel setText:@"-"];
+    [self.armedLabel setTextColor:[UIColor whiteColor]];
+
+    [self.customModeLabel setText:@"-"];
+    [self.customModeLabel setTextColor:[UIColor whiteColor]];
```
Comment (cnatoli) [inline iGCS/UserInterface/ViewControllers/GCSMapViewController.m]: customModeLabel color is always white afaics

File: iGCS/Models/Craft/GCSCraftArduPlane.m [hunk 11]
```diff
@@ -30,11 +36,34 @@ @implementation GCSCraftArduPlane
...
+                                         usingBlock:^(NSNotification *note) {
+                                             [weakSelf clearHearbeat];
+                                         }];
+
+        // set heartbeat to nil when app loses telemetry for a period of time
+        [self.notificationCenter addObserverForName:GCSHeartbeatManagerHeartbeatWasLost
+                                             object:nil
+                                              queue:self.craftQueue
+                                         usingBlock:^(NSNotification *note) {
+                                             [weakSelf clearHearbeat];
```
Comment (cnatoli) [inline iGCS/Models/Craft/GCSCraftArduPlane.m]: Unclear we want to do this:
* when telemetry is re-established, will speak the various alerts (regardless of whether or not it changed since last heartbeat was received)
* potentially invalidates the values of isInAutoMode etc

File: iGCS.xcodeproj/xcshareddata/xcschemes/iGCS.xcscheme [hunk 12]
```diff
@@ -68,7 +68,7 @@
       selectedLauncherIdentifier = "Xcode.DebuggerFoundation.Launcher.LLDB"
       launchStyle = "0"
       useCustomWorkingDirectory = "NO"
-      buildConfiguration = "Release"
+      buildConfiguration = "Debug"
```
Comment (cnatoli) [inline iGCS.xcodeproj/xcshareddata/xcschemes/iGCS.xcscheme]: Intended?

**Comment by cnatoli** [comment]
Assuming comments are addressed, change looks solid.

With an emphasis on *looks*. Feels like a bit of a violation of our process to push this as a release fix - istm we should go with a much less impactful change for 1.2, and push this to 1.3.

**Comment by andymoe** [comment]
Thanks for the review @cnatoli 

I agree this "fix" got way out of hand in terms of complexity once I tried to address all the edge cases but I wanted to push it to something closer to completion instead of abandoning it. I think you are right that I will introduce more risk at this stage than just doing the very minimal fix we spoke about in the first place.

I'll create a new branch from release-1-2-0 this afternoon and have you and @btgalush take a look at the much simpler approach that does not address all additional edge cases since those look a lot more like features than edge cases at this point. Thanks again for getting me a review last night.

**Comment by btgalush** [comment]
I'll test it with a controller this afternoon.

Sent from my iPhone

On Feb 22, 2015, at 8:42 AM, Andy Brown <notifications@github.com> wrote:

Thanks for the review @cnatoli <https://github.com/cnatoli>

I agree this "fix" got way out of hand in terms of complexity once I tried
to address all the edge cases but I wanted to push it to something closer
to completion instead of abandoning it. I think you are right that I will
introduce more risk at this stage than just doing the very minimal fix we
spoke about in the first place.

I'll create a new branch from release-1-2-0 this afternoon and have you and
@btgalush <https://github.com/btgalush> take a look at the much simpler
approach that does not address all additional edge cases since those look a
lot more like features than edge cases at this point. Thanks again for
getting me a review last night.

—
Reply to this email directly or view it on GitHub
<https://github.com/fightingwalrus/iGCS/pull/184#issuecomment-75444158>.

**Comment by btgalush** [comment]
Tested and found no issues.

I really prefer this functionality of going to a "-" for ARMED status when
returning from background versus displaying the stale data as the
"fix-mav-custom-mode-status-regression" branch does when there is no
telemetry available.  Not this release, but in the future I would like the
ARMED status to go to "-" when we lose telemetry.  If we don't know we
should not display the ARMED status one way or the other.  The other stale
data is acceptable and useful (last known battery voltage), but I don't
like the ever displaying a false armed status even with large "Telemetry
lost" text displayed.
